File: bindings/swift/Sources/Proven/SafePhone.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Phone number parsing delegated to libproven FFI.
///
/// E.164 phone number parsing and formatting via the formally
/// verified Idris 2 core.

import CProven

/// Parsed phone number.
public struct PhoneNumber: Equatable, Sendable {
    public let countryCode: UInt16
    public let nationalNumber: UInt64
}

public enum SafePhone {
    /// Parse a phone number from string.
    public static func parse(_ input: String) -> Result<PhoneNumber, ProvenError> {
        withStringBytes(input) { ptr, len in
            let result = proven_phone_parse(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(PhoneNumber(
                countryCode: result.country_code,
                nationalNumber: result.national_number
            ))
        }
    }

    /// Format a phone number in E.164 format (e.g., "+15551234567").
    public static func formatE164(countryCode: UInt16, nationalNumber: UInt64) -> Result<String, ProvenError> {
        consumeStringResult(proven_phone_format_e164(countryCode, nationalNumber))
    }

    /// Format a parsed phone number in E.164 format.
    public static func formatE164(_ phone: PhoneNumber) -> Result<String, ProvenError> {
        formatE164(countryCode: phone.countryCode, nationalNumber: phone.nationalNumber)
    }
}

File: bindings/prolog/safe_phone.pl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafePhone - FFI bindings to libproven phone number operations.
%% All phone number parsing is performed in verified Idris 2 code via libproven.

:- module(safe_phone, [
    parse_phone/2,
    format_phone_e164/3,
    is_valid_phone/1
]).

:- use_foreign_library(foreign(libproven)).

%% FFI declarations for libproven phone functions.
:- foreign(proven_phone_parse_ffi, c,
           proven_phone_parse(+string, +integer,
                              [-integer], [-integer], [-integer], [-integer])).
:- foreign(proven_phone_format_e164_ffi, c,
           proven_phone_format_e164(+integer, +integer,
                                    [-integer], [-string])).

%! parse_phone(+Input, -Result) is det.
%
%  Parse phone number to E.164 format via libproven.
%  Result is ok(phone(CountryCode, NationalNumber)) or error(parse_failure).
parse_phone(Input, Result) :-
    atom_string(Input, Str),
    atom_length(Input, Len),
    proven_phone_parse_ffi(Str, Len, Status, CountryCode, NationalNumber, IsValid),
    (   Status =:= 0, IsValid =:= 1
    ->  Result = ok(phone(CountryCode, NationalNumber))
    ;   Result = error(parse_failure)
    ).

%! format_phone_e164(+CountryCode, +NationalNumber, -Formatted) is det.
%
%  Format phone number as E.164 string via libproven.
format_phone_e164(CountryCode, NationalNumber, Formatted) :-
    proven_phone_format_e164_ffi(CountryCode, NationalNumber, Status, ResultStr),
    (   Status =:= 0
    ->  atom_string(Formatted, ResultStr)
    ;   Formatted = ''
    ).

%! is_valid_phone(+Input) is semidet.
%
%  Check if input is a valid phone number via libproven.
is_valid_phone(Input) :-
    parse_phone(Input, ok(_)).
